### DISCUSSION TITLE
Add JUnit tests to releaseNote

### DISCUSSION THREAD

**Discussion by aliakhlaghi93** [rfennell/AzurePipelines#1045] [Q&A]
Hello, 
I am trying to add my JUnit test to the release note. Tests are published via [Publish Test Result Task](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/test/publish-test-results?view=azure-devops&tabs=trx%2Cyaml). releaseTest object is empty in that case. Do you know how it is possible that I include them in the object?
Or is there any way that I can define a new custom object?

Regards,
Ali

**Comment by rfennell** [comment]
As long as you have published your JUnit test results to the Azure DevOps pipeline they should appear. 

The first check is to confirm your test results appear in the Azure DevOps pipeline run test results tabs. If they are not there, my task cannot report on them, as all I do is programmatically access the data that populates that tab.

Also, check the detailed pipeline log to make sure of what is actually being published 

To my knowledge, there is no reason JUnit test should not be listed.

**Reply by aliakhlaghi93** [reply]
Thank you for your reply. I can see the result in my test tab and also publish test results task log shows that it is published successfully. Even I have test metaData in my env variables of that stage.
![image](https://user-images.githubusercontent.com/64731820/126333410-65d8a57c-21c4-44eb-84be-2ec27b74acfb.png)

I am trying to access the object in the release pipeline in the last stage for all previous stages. 

My current solution:  I am using a PowerShell script to parse the publish result log file and get runId of the test. Then get the test results from API and hard code all the values in my template. How do you get the test result? Do you use runId?

**Reply by rfennell** [reply]
I would expect the block in the Handlebars template for my Release Note Tasks to be similar to the following

```
# Global list of test ({{tests.length}})
{{#forEach tests}}
{{#if isFirst}}### Tests {{/if}}
* ** ID{{this.id}}** 
   -  Name: {{this.testCase.name}}
   -  Outcome: {{this.outcome}}
{{/forEach}}
```

You can of course dump all the available data using the following to get an idea of all the fields available 

```
{{ json tests}}
```

**Reply by aliakhlaghi93** [reply]
Yeah, My tests object contains tests that are executed in build. However, `releaseTest` is empty while I can see the result in the test tab of the release stage. My main question is how do you get the releaseTest object in the backend? So I can investigate why it becomes empty at the end.

**Reply by rfennell** [reply]
The code to get the release tests (and remember these are only added in Classic Releases, not Muti-Stage YAML Pipelines where the test results are in the global tests) is in [ReleaseNoteFunctions.ts](https://github.com/rfennell/AzurePipelines/blob/5d8a5ad72782dda8702f40c12f37f1be7bb11ab7/Extensions/XplatGenerateReleaseNotes/V3/ReleaseNotesFunctions.ts#L668)

```
export async function getTestsForRelease(
    testAPI: TestApi,
    teamProject: string,
    release: Release
): Promise<TestCaseResult[]> {
    return new Promise<TestCaseResult[]>(async (resolve, reject) => {
        let testList: TestCaseResult[] = [];
        try {
            for (let envIndex = 0; envIndex < release.environments.length; envIndex++) {
                const env = release.environments[envIndex];
                    let envTestResults = await (testAPI.getTestResultDetailsForRelease(teamProject, release.id, env.id));
                    if (envTestResults.resultsForGroup.length > 0) {
                        for (let index = 0; index < envTestResults.resultsForGroup[0].results.length; index++) {
                            const test =  envTestResults.resultsForGroup[0].results[index];
                            if (testList.filter(e => e.testRun.id === `${test.testRun.id}`).length === 0) {
                                tl.debug(`Adding tests for test run ${test.testRun.id}`);
                                let run = await (testAPI.getTestResults(teamProject, parseInt(test.testRun.id)));
                                testList.push(...run);
                            } else {
                                tl.debug(`Skipping adding tests for test run ${test.testRun.id} as already added`);
                            }
                        }
                    } else {
                        tl.debug(`No tests associated with release ${release.id} environment ${env.name}`);
                    }
            }
            resolve(testList);
        } catch (err) {
            reject(err);
        }
    });
}
```

So the call is to `getTestResults(<project>,<testtrun>)`

**Reply by aliakhlaghi93** [reply]
Haha, I found my mistake. An "s" should be added to the documentation.
![image](https://user-images.githubusercontent.com/64731820/126488107-90aad907-54ef-46c6-ac29-19a3b147769f.png)

**Reply by rfennell** [reply]
Thanks for pointing that out. I have updated the [WIKI](https://github.com/rfennell/AzurePipelines/wiki/GenerateReleaseNotes---Node-based-Cross-Platform-Task) with the correction and the next release of the extension will have an updated readme (and hence marketplace entry)
